app/modules/file.c

```diff
@@ -214,18 +214,60 @@ static int file_open( lua_State* L )
 static int file_list( lua_State* L )
 {
   vfs_dir  *dir;
+  const char *pattern;
+  struct vfs_stat stat;
+  int pcres;
+  int res;
+
+  lua_settop(L, 1);
+  pattern = luaL_optstring(L, 1, NULL);   /* Pattern (arg) or nil (not) at 1 */
+
+  dir = vfs_opendir("");
+  if (dir == NULL) {
+    return 0;
+  }
+
+  lua_newtable( L );                      /* Table at 2 */
+
+  if (pattern) {
+    /* 
+     * We know that pattern is a string, and so the "match" method will always
+     * exist.  No need to check return value here
+     */
+    luaL_getmetafield( L, 1, "match" );  /* Function at 3 */
+  }
 
-  if (dir = vfs_opendir("")) {
-    lua_newtable( L );
-    struct vfs_stat stat;
-    while (vfs_readdir(dir, &stat) == VFS_RES_OK) {
-      lua_pushinteger(L, stat.size);
-      lua_setfield(L, -2, stat.name);
+  while (vfs_readdir(dir, &stat) == VFS_RES_OK) {
+    if (pattern) {
+      lua_settop( L, 3 );                 /* Ensure nothing else on stack */
+
+      /* Construct and pcall(string.match,name,pattern) */
+      lua_pushvalue( L, 3 );
+      lua_pushstring( L, stat.name );
+      lua_pushvalue( L, 1 );
+      pcres = lua_pcall( L, 2, 1, 0 );
```

Comment:
This is really the wrong thing to do.  You should just use a `lua_call()` and let the lua error system do its stuff.  So for example if you do something like:
 ```Lua
    for name, len in pairs(files.list('[a-f]%')) do`
```
you want it  to throw a `malformed pattern (ends with '%')` rather that a `bad argument #1 to 'pairs' (table expected, got number)`.  9 times out of 10 the pattern is programmed and on the 10th time, the programmer can always wrap the `file.list(filter)` in a `pcall()`

Reply:
As I said above, I do not believe that `longjump`-ing out of `file.list()` is correct, as that will not call `vfs_closedir()` on its way out.  I'll rework the patch to use `lua_error` to re-throw the exception.